### ISSUE TITLE
api-platform as framework agnostic solution

### ISSUE THREAD

**Issue by guilhermeaiolfi** [api-platform/api-platform#713]
Hey,

I really like the idea behind this project. But I'm not using Symphony and I don't intend to. I want to integrate api-platform in an existing project that uses Zend Expressive. I read all the docs about configuration and data providers. But all of them seems to use Symphony one way or another. 

I specially interested in exposing a JSONAPI and a GraphQL interface for my client app. And I already have my models as Doctrine Entities (no schema.org references). So, I was wondering if it's possible to do what I want without Symphony at all? How? 

Thanks.

**Comment by teohhanhui** [comment]
Please let us know what are the pain points when using API Platform outside of the Symfony framework. I believe it's an ongoing goal for us to be framework agnostic.

**Comment by soyuka** [comment]
I think that someone already worked on using Laravel with Api Platform and it is definitely in our scope to make it work with other frameworks!

**Comment by guilhermeaiolfi** [comment]
I realized I was looking for similar issues in the wrong repository. I discovered api-platform today, so I don't know how it works. But for what I've seen so far, moving away from ```http-foundation``` and embracing PSR-7 would a pretty good start. It was discussed here: https://github.com/api-platform/core/issues/474, but it was closed without any explanation. Maybe we should close this issue and reopen that one? Same thing happened here: https://github.com/api-platform/core/issues/474

**Comment by teohhanhui** [comment]
You might be able to use the Symfony PSR-7 Bridge: https://symfony.com/doc/current/components/psr7.html

**Comment by dunglas** [comment]
Hi,

Indeed, API Platform v2 has been designed to be framework agnostic, and supporting other frameworks than Symfony such as Laravel, and PSR-7 / PSR-15 is definitely something we want.

Things have improved since we added GraphQL support (the code base has been refactored, and all dependencies to Symfony's HttpFoundation have been removed from the Doctrine Bridge).

As you pointed out, the main coupling between API Platform and Symfony full-stack are the event listener. We are in the process of removing as much logic as possible from event listeners and move them in dedicated classes (and services). It's not totally finished, but most of the work has been done.
The listeners and the few controllers now should just extract data from Symfony's Request or add some data to the Symfony's Response. That's all.

To implement PSR-7 / PSR-15, we'll have to create new PSR-15 middlewares for every existing listener/controller. Those middlewares will call the exact same services than the event listeners. It shouldn't be very hard, but it's some work. And it's definitely something we want in the main project.

Except middlewares, here are the other things to do to have full Expressive support:

* registering the classes in Zend Service Manager (except it Expressive can work with the Symfony DI container, I don't know Expressive enough)
* Implementing our `IriConverterInterface` and [some related interfaces](https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Routing/IriConverter.php) using Expressive's router (except it Expressive can work with the Symfony Router, I don't know Expressive enough)

It's some work, but I think it will not be that hard in the end. If you start working on this, we definitely will merge it in the main repository, and I'll be glad to help. In my opinion we should focus on supporting PSR-7 / PSR-15, not Expressive specifics. I mean if it's possible to use SF's DI and Router, we should go this way, it will make the initial development easier and reduce the maintenance.

Using the HttpFoundation <-> PSR-7 bridge (that I've created) isn't a good idea in API Platform, because it comes with an important performance penalty.

**Comment by teohhanhui** [comment]
I wonder if it's possible to support PSR-11 container or is that not expressive (😎) enough?

**Comment by dunglas** [comment]
> I wonder if it's possible to support PSR-11 container or is that not expressive (😎) enough?

We already do (because the Symfony container implements PSR-11). The problem is: PSR-11 only allows to retrieve services from the container, not how to register them.

**Comment by teohhanhui** [comment]
Yeah, I realized that. So it means the user can choose to use the Symfony DI container, which we integrate with out of the box, or they could choose to use their own DI container but they'll have to register the services by themselves. I don't see any other way to do it for now...

**Comment by guilhermeaiolfi** [comment]
GraphQL format is still using HttpFoundation for request and responses (besides listeners). 

If we are going to the framework agnostic road, we should consider api-platform as a library and not a framework. And it being a library I don't think it should be using DI containers at all. 

What I would like to see is an example of using api-platform without routes, DI containers, listeners or any other concept that belongs to the framework scope. Basically take a request, read its content, configure the schema and make the library respond to what's been asked for in the format I want. I was able to do that in GraphQL in 60 lines of code. 

I may be wrong, but api-platform seems to solve the same problem with an extra layer of abstraction. So the thing I did using GraphQL should be an achievable goal too and it would reduce the complexity to get started.

Middlewares/listeners/services would exist to automate the integration to frameworks but not a hard dependency to use the library.

**Comment by dunglas** [comment]
> GraphQL format is still using HttpFoundation for request and responses (besides listeners).

Only in the controller (it's mandatory), but not in the lower level layer.

> If we are going to the framework agnostic road, we should consider api-platform as a library and not a framework. And it being a library I don't think it should be using DI containers at all.

It's both a set of libraries and a framework (similar to SF components / SF full stack). We follow use a monolithic repository, but we'll create a subtree split at some point.
Libraries need to be registered in DI containers (most libraries have related bundles for Symfony, modules for Laravel). It's why the DI plumbing is in the `Bridge` namespace. We can add more adapters to more frameworks/DI.

> What I would like to see is an example of using api-platform without routes, DI containers, listeners or any other concept that belongs to the framework scope. Basically take a request, read its content, configure the schema and make the library respond to what's been asked for in the format I want.

I would be great to have a doc entry like this, I agree. Maybe @gorghoa do you have some hints about this?

> I was able to do that in GraphQL in 60 lines of code.

You mean using `webonyx/graphql-php`?

> I may be wrong, but api-platform seems to solve the same problem with an extra layer of abstraction. 

Indeed.

> So the thing I did using GraphQL should be an achievable goal too and it would reduce the complexity to get started.

I agree!

> Middlewares/listeners/services would exist to automate the integration to frameworks but not a hard dependency to use the library.

Exactly.

**Comment by gorghoa** [comment]
> I would be great to have a doc entry like this, I agree. Maybe @gorghoa do you have some hints about this?

Thanks @dunglas  for inviting me to this discussion :)

Unfortunately, I did not took that path all the way down yet. Given the timebox I had, I simply wrote some custom data provider/persister targeting custom domain boundaries and « proxified » all the api routes  to a standard api-platform instance to handle routing/content negociation/security and the like.

I do not regret it, since I was able to leverage some symfony niceties (especially DI and security).  I doubt that this is an option for you @guilhermeaiolfi though.

**Comment by 3kynox** [comment]
Hello there.

I read someone accomplished running api-platform on Laravel.
Is there any location that explain the process ?

Thanks in advance

**Comment by guilhermeaiolfi** [comment]
Hello again, good people,

any work was done to make it easier to accomplished this?

**Comment by soyuka** [comment]
Nothing has been made AFAIK. 

First step would be to implement the Laravel Dependency Injection system for our classes (provider?). Then register the routes, just like we're doing with Symfony in the `Bridge/Symfony/Routing` namespace. 
Then the event listeners system. 

That's a lot of work though but it's definitely feasible in a few weeks! I'd really love to do this but I've nor the time nor the money to :crying_cat_face:. 

Eloquent may come later (also https://github.com/theofidry/EloquentSerializer).

**Comment by rastapopougros** [comment]
Hello, 4 years later, is there an update on this topic?

Same use case for me, all the functionalities are really exactly what I need to add APIs to *existing sites that are not made with symfony*, and sometimes even not with any specific framework (like with some CMS for ex).

Imagine if the "entry" URL `site.tld/platform.api` and all sub URLs `site.tld/platform.api/any/thing?else` are managed by some PHP class or method (by this or that CMS or else)… so I just wish that in this personal code, I could call some kind of `\ApiPlatform->run(…)` providing the relevant part of the URL (here `/any/thing?else`) and, go go, API-Platform would make the magic, use that as the real endpoint and optional parameters to compile and send the final JSON or else.

**Comment by dunglas** [comment]
Yes! See https://api-platform.com/docs/core/bootstrap/

It would be even better with API Platform 2.7 (available in alpha).

**Comment by rastapopougros** [comment]
Oh thanks for this answer!
So, if I understand well, I "just" copypaste all this code in the file/function that manage my `site.tld/entrypoint.api/xxx` URLs?

I have so much questions about the non-framework usage… :p 
Like for filters, I read
> If you use [custom data providers](https://api-platform.com/docs/core/data-providers/), you can still create filters by implementing the previously mentioned interface, but - as API Platform isn't aware of your persistence system's internals - you have to create the filtering logic by yourself.

but it's not totally clear how easy or not it is cause all examples are for adding ORM filters, not from a custom provider :)

**Comment by rastapopougros** [comment]
Maybe @soyuka also could give some comments and tips, as he seems to be the parent of those… a bit rough and tough hundreds of lines :) 
https://gist.github.com/soyuka/38863de8374eabb9bf0a489ca10a797b

**Comment by rastapopougros** [comment]
@dunglas thanks again for the path to follow, but I would suggest you to not publish this code in your "docs" section without a *clear* introduction saying it's old and don't run "as is". There are fatal errors at least, and no comments or instructions to follow to understand which part we must modify/configure, etc. It's really rough and tough, this long (now buggy) code without anything around to explain :) 

I made some questions and remarks below the original gist, too, if it can help to clarify the use.

Anyway, it's still a goal to be able to use this lib in a really autonomous way…

**Comment by soyuka** [comment]
Closing per #2987. Questions are best asked on [GitHub Discussions](https://github.com/api-platform/core/discussions) or [Stack Overflow (api-platform tag)](https://stackoverflow.com/questions/tagged/api-platform.com). The framework now lives at `api-platform/core`.
